library/bignum.c

```diff
@@ -1877,19 +1877,31 @@ int mbedtls_mpi_mod_int( mbedtls_mpi_uint *r, const mbedtls_mpi *A, mbedtls_mpi_
 
 /*
  * Fast Montgomery initialization (thanks to Tom St Denis)
+ *
+ * This computes the negative modular inverse of m in mbedtls_mpi_uint.
  */
-static void mpi_montg_init( mbedtls_mpi_uint *mm, const mbedtls_mpi *N )
-{
-    mbedtls_mpi_uint x, m0 = N->p[0];
-    unsigned int i;
-
-    x  = m0;
-    x += ( ( m0 + 2 ) & 4 ) << 1;
-
-    for( i = biL; i >= 8; i /= 2 )
-        x *= ( 2 - ( m0 * x ) );
-
-    *mm = ~x + 1;
+static void mpi_montg_init( mbedtls_mpi_uint *m_inv,
+                            mbedtls_mpi_uint const *m )
+{
+    /* The basic principle is the following:
+     * If R=2^k and we already know that m_inv is the negative
```

Comment:
I'm having trouble following this explanation. What's R? What's k? What if R ≠ 2^k?

Reading the whole thing and trying to put it in context, I guess you mean we're interested in the problem of calculating a modular inverse modulo 2^k for some nonnegative integer k, and we define R as shorthand for 2^k? If so, please rewrite to make this clear. Something like:

> Let k >= 0 and R = 2^k. We want to calculate the inverse of m modulo R, that is m_inv such that …

Also I think it would be clearer to say upfront that we assume that m is odd.